Automattic/jetpack
# [Boost] Add CSS/JS minification

## Proposed changes:

* Add all [Page Optimize](https://github.com/automattic/page-optimize) features under a new feature in Boost - Minify.
* Feature can be toggled on or off from the Boost UI.
**Note:** The feature is called Minify, but only supports what Page Optimize supports - concatenation for CSS and JS. It also supports minification of CSS, but that gets enabled with a PHP constant.
![image](https://user-images.githubusercontent.com/11799079/217846550-2d3252bb-fc1a-4392-ad8b-f5683c896173.png)
* Page Optimize had a settings page under `Settings -> Performance`. That has been removed. There's a discussion around the UI (pc9hqz-1zN-p2) about how we're going to handle the settings transition.
* For the time being, existing settings from Page Optimize get migrated with new names. Migration is not ideal, should be changed later.
* Added GET param (`jb-disable-minify`) to disable Minify to make testing easier.
* The Minify feature and the Page Optimize plugin can be active at the same time, though it's not advised, as they share the same code and it will get executed twice.
* Since Page Optimize relied on PHP constants, code has been written with the idea, that users might have updated settings and this feature [aims to respect them](https://github.com/Automattic/jetpack/blob/5ffc0432043ed166f7f3e2f6d0daa166896524a5/projects/plugins/boost/app/features/optimizations/minify/Config.php).

### Other information:


## Jetpack product discussion



pc9hqz-1yh-p2

## Does this pull request change what data or activity we track or use?



No.

## Testing instructions:






1. Install this version of Boost;
2. Allow the feature on the website, by adding this to `wp-config.php` - `define( 'JETPACK_BOOST_MINIFY', true );`;
3. Enable the Minify feature;
4. Open a page and check source for `_static`. This is the URL the feature uses when serving its cache;
![image](https://user-images.githubusercontent.com/11799079/217850333-2bede350-0fc5-44e6-b637-78b301157df8.png)

5. Enable `WP_DEBUG` to see the handles of the scripts/styles included in the cached url.
![image](https://user-images.githubusercontent.com/11799079/217850116-a501c28b-34c4-4273-8a36-7ddca3b5f311.png)
You might also notice HTML comments - that's the debug output from the feature (shows up only if `WP_DEBUG` is set to `true`).
![image](https://user-images.githubusercontent.com/11799079/217851726-e972cb8a-4ba2-4bb6-bfc6-510ecac752a4.png)
